pkg/kp/statusstore/consul_store.go

```diff
@@ -0,0 +1,161 @@
+}
+
+var _ Store = &consulStore{}
+
+func NewConsul(client *api.Client) Store {
+	return &consulStore{
+		kv: client.KV(),
+	}
+}
+
+type noStatusError struct {
+	key string
+}
+
+func (n noStatusError) Error() string {
+	return fmt.Sprintf("No status record found at %s", n.key)
+}
+
+func IsNoStatus(err error) bool {
+	_, ok := err.(noStatusError)
+	return ok
+}
+
+func (s *consulStore) SetStatus(t ResourceType, id string, namespace Namespace, status Status) error {
+	key := namespacedResourcePath(t, id, namespace)
+	pair := &api.KVPair{
+		Key:   key,
+		Value: status.Bytes(),
+	}
+	_, err := s.kv.Put(pair, nil)
+	if err != nil {
+		return consulutil.NewKVError("put", key, err)
+	}
+
+	return nil
+}
+
+func (s *consulStore) GetStatus(t ResourceType, id string, namespace Namespace) (Status, error) {
+	key := namespacedResourcePath(t, id, namespace)
+	pair, _, err := s.kv.Get(key, nil)
+	if err != nil {
+		return nil, consulutil.NewKVError("get", key, err)
+	}
+
+	if pair == nil {
+		return nil, noStatusError{key}
+	}
+
+	return pair.Value, nil
+}
+
+func (s *consulStore) DeleteStatus(t ResourceType, id string, namespace Namespace) error {
+	key := namespacedResourcePath(t, id, namespace)
+	_, err := s.kv.Delete(key, nil)
+	if err != nil {
+		return consulutil.NewKVError("delete", key, err)
+	}
+
+	return nil
+}
+
+func (s *consulStore) GetAllStatusForResource(t ResourceType, id string) (map[Namespace]Status, error) {
+	ret := make(map[Namespace]Status)
+	prefix := resourcePath(t, id)
+	pairs, _, err := s.kv.List(prefix, nil)
+	if err != nil {
+		return nil, consulutil.NewKVError("list", prefix, err)
+	}
+
+	for _, pair := range pairs {
+		key := pair.Key
+		status := pair.Value
+
+		_, _, namespace, err := keyParts(key)
+		if err != nil {
+			return nil, err
+		}
+
+		ret[namespace] = status
+	}
+
+	return ret, nil
+}
+
+func (s *consulStore) GetAllStatusForResourceType(t ResourceType) (map[string]map[Namespace]Status, error) {
+	ret := make(map[string]map[Namespace]Status)
+	prefix := resourceTypePath(t)
+	pairs, _, err := s.kv.List(prefix, nil)
+	if err != nil {
+		return nil, consulutil.NewKVError("list", prefix, err)
+	}
+
+	for _, pair := range pairs {
+		key := pair.Key
+		status := pair.Value
+
+		_, id, namespace, err := keyParts(key)
+		if err != nil {
+			return nil, err
+		}
+
+		_, ok := ret[id]
+		if !ok {
+			ret[id] = make(map[Namespace]Status)
+		}
+
+		ret[id][namespace] = status
+	}
+
+	return ret, nil
+}
+
+func resourceTypePath(t ResourceType) string {
+	return path.Join(statusTree, t.String())
+}
+
+func resourcePath(t ResourceType, id string) string {
+	return path.Join(resourceTypePath(t), id)
+}
+
+func namespacedResourcePath(t ResourceType, id string, namespace Namespace) string {
+	return path.Join(resourcePath(t, id), namespace.String())
+}
+
+// in consul, the key for a status looks like
+// '/status/pod_clusters/<pod_cluster_id>/<namespace>'.
+func keyParts(key string) (ResourceType, string, Namespace, error) {
```

Comment:
because ResourceType might not be pod cluster. I suppose I could make it be some generic ID but that would probably be weird for callers? e.g. `statusstore.ID(podId)`